transport/peer/peerlist/roundrobin.go

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package peerlist
+
+import (
+	"context"
+
+	"go.uber.org/yarpc/transport"
+	"go.uber.org/yarpc/transport/internal/errors"
+
+	"github.com/uber-go/atomic"
+)
+
+// NewRoundRobin creates a new round robin PeerList using
+func NewRoundRobin(peerIDs []transport.PeerIdentifier, agent transport.Agent) (*RoundRobin, error) {
+	rr := &RoundRobin{
+		pr:    NewPeerRing(len(peerIDs)),
+		agent: agent,
+	}
+
+	err := rr.addMulti(peerIDs)
+	return rr, err
+}
+
+// RoundRobin is a PeerList which rotates which peers are to be selected in a circle
+type RoundRobin struct {
+	pr      *PeerRing
+	agent   transport.Agent
+	started atomic.Bool
+}
+
+func (pl *RoundRobin) addMulti(peerIDs []transport.PeerIdentifier) error {
+	errs := errors.Errors{}
+
+	for _, peerID := range peerIDs {
+		p, err := pl.agent.RetainPeer(peerID, pl)
```

Comment:
perhaps validate up front and return an error for any peer identifier without a corresponding peer, signaling that the entire operation failed without effect (instead of partially succeeding).